src/v/ssx/async_algorithm.h

```diff
@@ -106,6 +106,10 @@ template<std::random_access_iterator I, typename Fn>
 iter_size<I> for_each_limit(const I begin, const I end, ssize_t limit, Fn f) {
     auto chunk_size = std::min(limit, end - begin);
     I chunk_end = begin + chunk_size;
+    // std::for_each discards [[nodiscard]] callable result, check explicitly
+    if (false) {
```

Comment:
would it be possible to check if function passed to for each limit returns void by using a cpp concept ?

Reply:
there is another flavor of this function which isn't validated

Reply:
> would it be possible to check if function passed to for each limit returns void by using a cpp concept?

Probably, but I think this would be too strict. One may want to use a function that actually returns a value which is by design okay to discard.

> there is another flavor of this function which isn't validated

The one below? It is not affected by the problem, as `f` is called explicitly. If `f` or its return type is marked `[[nodiscard]]` the call will emit warning (as error) on compilation, I have checked. `future<...>` class is declared `[[nodiscard]]` in Seastar.

Reply:
Does this even work? 

It looks like the moment you convert the function to a function pointer the no discard attribute gets lost https://gcc.godbolt.org/z/jWbKo59r8

Reply:
Yes it does work. Nodiscard attribute gets lost from a function if you convert it to a pointer indeed, as it's stored neither in the pointer nor in the pointer type. It is not, however, lost if the return type itself is marked as nodiscard: https://gcc.godbolt.org/z/bGq1ordvj. In this case the function pointer type retains the information about the function return type, which is known to be nodiscard.